<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>onerror</title>
</head>
<body>
    
    <h3>onerror event</h3>
    <p>
        今天我所负责项目的一个页面在某个机型上出现了个兼容问题，代码已经是趋于稳定的节奏忽然又出现这个问题，并且还是特定机型能复现，难免感觉有点奇葩，但问题是app内嵌页面在手机上的调试，因为m站均无问题，就导致一些后台报错无法直观的在手机上看到，这样情况下的调试难免有点盲人摸象的感觉。
        要捕获页面的错误，首先想到的肯定是try-catch，虽然问题还没有得到解决，但是感觉这个技能还是有必要掌握一下的，同时也发现要系统的学习下JS的报错机制以及报错类型和常见错误对应的处理方式的必要性。
    </p>

    <script>
        
        /*
        *   全局一旦有js脚本报错，便会触发onerror事件
        *
        *   1、若处理函数返回true，则浏览器不会弹出错误信息框，否则，浏览器会会弹出错误信息；
        *   2、只有运行错误才会出发onerror，语法错误不会触发（具体JS错误类型，会下一次具体描述）；
        *       具体是以下几种：
        *           运行时错误（runtime error），例如无效的对象引用或安全限制或主动抛出的错误
                    下载错误（fail error），如图片
                    在IE9中，获取多媒体数据失败也会引发
                    或者代码存在语法错误，导致编译不通过时也会引发
            3、当执行的js脚本存在跨域情况时，给返回头部添加Access-Control-Allow-Origin: *属性，给js标签添加crossorigin 属性
            
        *   参考链接：https://blog.sentry.io/2016/01/04/client-javascript-reporting-window-onerror.html
        *            https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onerror
        * */
        
        window.onerror = function (msg, url, lineNo, columnNo, error) {
            var string = msg.toLowerCase();
            var substring = "script error";
            if (string.indexOf(substring) > -1){
                alert('Script Error: See Browser Console for Detail');
            } else {
                var message = [
                    'Message: ' + msg,
                    'URL: ' + url,
                    'Line: ' + lineNo,
                    'Column: ' + columnNo,
                    'Error object: ' + JSON.stringify(error)
                ].join(' - ');

                alert(message);
            }

            var suppressErrorAlert = false;
//            suppressErrorAlert = true;

            return suppressErrorAlert;
        };
        
        var pElem = document.getElementsByTagName("p")[0];
        var showInfo = function () {
            call_undefined_fun();
        };
        pElem.addEventListener("click", showInfo);
        
        
        /*
        * JS错误处理机制知识点总结：
        *
        *   1、Error对象
        *    JS解析或执行时，一旦发生错误，就会抛出错误对象，所有的对象都是Error的实例，每个实例都具有三个属性：
        *       message：错误信息，也即传给Error("info")对象的参数"info"；
        *       name：错误名字，默认为"Error"，它是一个可写的属性；
        *       stack： 错误的堆栈，由最底层到最外层地存放错误的堆栈信息，据此我们可以知道错误是如何一级一级传递的以及错误的源头；
        *   2、JS的错误类型
        *   除了Error这种错误类型，JS还派生除了其他六种错误类型（最常见的为前三种）：
        *       SyntacError：解析代码时发生的语法错误；
        *       ReferanceError：引用一个不存在的变量时发生的错误；
        *       TypeError： 变量或参数不是预期类型；
        *       RangeError：变量值超出了有效范围；
         *      URIError：URI相关的函数参数不正确时抛出的错误；
        *       EvalError：eval函数没有被正确执行；
        *   3、throw语句 中断语句的执行，我们可以用其来抛出一个异常，参数可以为各种类型的值，可以是一个Error对象，或自定义的错误对象；
        *   4、try-catch、try-catch-finally语句，用法与其他面向对象的语言比较类似；
        * */
        
        
        function assertion(expression, msg) {
            if (!expression) {
                throw {name : "Assertion Exception", message : msg}
            }
        }
        
        assertion(typeof param === "i am param", "param is not defined~");
        
        //上述方法等同于：
        console.assert(typeof param === "i am param", "param is not defined~");
        
        
    </script>
</body>
</html>














